pkg/datapath/tables/node_address.go

```diff
@@ -0,0 +1,365 @@
+				devices, _ := n.Devices.All(txn)
+				for dev, _, ok := devices.Next(); ok; dev, _, ok = devices.Next() {
+					n.update(txn, nil, n.getAddressesFromDevice(dev), nil)
+				}
+				txn.Commit()
+
+				// Start the job in the background to incremental refresh
+				// the node addresses.
+				return g.Start(ctx)
+			},
+			OnStop: g.Stop,
+		})
+
+}
+
+func (n *nodeAddressController) run(ctx context.Context, reporter cell.HealthReporter) error {
+	defer n.tracker.Close()
+
+	limiter := rate.NewLimiter(nodeAddressControllerMinInterval, 1)
+	revision := statedb.Revision(0)
+	for {
+		txn := n.DB.WriteTxn(n.NodeAddresses)
+		process := func(dev *Device, deleted bool, rev statedb.Revision) error {
+			addrIter, _ := n.NodeAddresses.Get(txn, NodeAddressDeviceNameIndex.Query(dev.Name))
+			existing := sets.New[NodeAddress](statedb.Collect[NodeAddress](addrIter)...)
+			var new sets.Set[NodeAddress]
+			if !deleted {
+				new = n.getAddressesFromDevice(dev)
+			}
+			n.update(txn, existing, new, reporter)
+			return nil
+		}
+		var watch <-chan struct{}
+		revision, watch, _ = n.tracker.Process(txn, revision, process)
+		txn.Commit()
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-watch:
+		}
+		if err := limiter.Wait(ctx); err != nil {
+			return err
+		}
+	}
+}
+
+func (n *nodeAddressController) update(txn statedb.WriteTxn, existing, new sets.Set[NodeAddress], reporter cell.HealthReporter) {
+	updated := false
+
+	// Insert new addresses that did not exist.
+	for addr := range new {
+		if !existing.Has(addr) {
+			updated = true
+			n.NodeAddresses.Insert(txn, addr)
+		}
+	}
+
+	// Remove addresses that were not part of the new set.
+	for addr := range existing {
+		if !new.Has(addr) {
+			updated = true
+			n.NodeAddresses.Delete(txn, addr)
+		}
+	}
+
+	if updated {
+		addrs := showAddresses(new)
+		n.Log.WithField("node-addresses", addrs).Info("Node addresses updated")
+		if reporter != nil {
+			reporter.OK(addrs)
+		}
+	}
+}
+
+func (n *nodeAddressController) getAddressesFromDevice(dev *Device) (addrs sets.Set[NodeAddress]) {
+	addrs = sets.New[NodeAddress]()
+
+	if dev.Flags&net.FlagUp == 0 {
+		return
+	}
+
+	// Skip obviously uninteresting devices.
+	// We include the HostDevice as its IP addresses are consider node addresses
+	// and added to e.g. ipcache as HOST_IDs.
+	if dev.Name != defaults.HostDevice {
+		skip := false
+		for _, prefix := range defaults.ExcludedDevicePrefixes {
+			if strings.HasPrefix(dev.Name, prefix) {
+				skip = true
+				break
+			}
+		}
+		if skip {
+			return
+		}
+	}
+
+	// ipv4 and ipv6 are set to true when the primary address is picked.
+	// Used to implement 'NodePort' and 'Primary' flags.
+	ipv4, ipv6 := false, false
+
+	for _, addr := range sortAddresses(dev.Addrs) {
+		// We keep the scope-based address filtering as was introduced
+		// in 080857bdedca67d58ec39f8f96c5f38b22f6dc0b.
+		if addr.Scope > uint8(n.AddressScopeMax) {
+			// Addresses are sorted by scope, so can stop early.
```

Comment:
but they're sorted by primary/secondary first, so a primary addr with larger scope would break but should not, afaict?